vignettes/invariants.Rmd

```diff
@@ -198,42 +192,147 @@ df[[TRUE]]
-df[[1.5]]
+`x$name` is equal to `x[["name"]]`.
+Unlike data frames, tibbles do not partially match names.
+
+```{r dollar-equivalent-to-subset-pmatch, dftbl = TRUE}
+df$l
 ```
 
-## Definition of `x$name`
+Because `df$x` is rarely used in packages, it can raise a warning:
 
-`x$name` is equal to `x[["name"]]`.
+```{r dollar-j-oob-character, dftbl = TRUE}
+df$not_present
+```
 
-<!-- I removed the examples here because I think the principle is so simple -->
+## Column subsetting
 
-Unlike data frames, tibbles do not partially match.
+### Definition of `x[j]`.
 
-```{r dollar-equivalent-to-subset-pmatch, dftbl = TRUE}
-df$l
+`j` is converted to an integer vector by `vectrs::vec_as_index(j, nrow(x), names = names(x))`. Then `x[c(j_1, j_2, ..., j_n)]` is equivalent to `tibble(x[[j_1]], x[[j_2]], ..., x[[j_3]])``.
+This implies that `j` must be a numeric or character vector, or a logical vector with length 1 or `nrow(x)`.
+
+```{r bracket-j-definition, dftbl = TRUE}
+df[1:2]
 ```
 
-## Definition of `x[[j]] <- a`.
+<!-- TODO: what should happen here? -->
+If values of `j` are repeated, then ???
+
+```{r backet-j-duplication, dftbl = TRUE}
+df[c(1, 1)]
+```
+
+`nrow(df[j])` equals `nrow(df)`. 
+
+```{r backet-j-empty, dftbl = TRUE}
+df[integer()]
+```
+
+### Definition of `x[, j]`
+
+`x[, j]` is equal to `x[j]`.
+Tibbles do not perform column extraction if `x[j]` would yield a single column.
+
+```{r bracket-drop, dftbl = TRUE}
+df[, 1]
+df[, 1:2]
+```
+
+### Definition of `x[, j, drop = TRUE]`
+
+For backward compatiblity, `x[, j, drop = TRUE]` performs column __extraction__, returning `x[, j][[1]]` when `ncol(x[, j])` is 1.
+
+```{r bracket-always-returns-tibble-drop, dftbl = TRUE}
+df[, 1, drop = TRUE]
+```
+
+## Row subsetting
+
+### Definition of `x[i, ]`
+
+`i` must be a numeric or logical vector of length `nrow(x)`, or a logical vector of length 1.
+
+```{r bracket-i-wrong-type, dftbl = TRUE}
+df[mean, ]
+df[list(1), ]
+```
+
+Unlike data frames, only logical vectors of length 1 are recycled.
+<!-- TODO: should this be an error? -->
+
+```{r bracket-i-recycle, dftbl = TRUE}
+df[c(TRUE, FALSE), ]
+```
+
+`i` is converted to an integer vector by `vectrs::vec_as_index(i, nrow(x))`.
+This implies that OOB values generate warnings:
+
+```{r bracket-i-oob, dftbl = TRUE}
+df[10, ]
+df["x", ]
+```
+
+`x[i, ]` is equivalent to `tibble(vec_slice(x[[1]], i), vec_slice(x[[2]], i), ...)`. 
+(It is not defined in terms of `x[[j]][i]` because that definition does not generalise to matrix and data frame columns.)
+
+NB: scalar logicals are recycled, but scalar numerics are not. That makes the `x[NA, ]` and `x[NA_integer_, ]` return different results.
+
+```{r bracket-i, dftbl = TRUE}
+df[NA, ]
+df[NA_integer_, ]
+```
+
+### Definition of `x[i, , drop = TRUE]`
+
+`drop = TRUE` has no effect when not selecting a single column:
+
+```{r backet-i-drop, dftbl = TRUE}
+df[1, , drop = TRUE]
+```
+
+<!-- TODO: I think this should probably be an error -->
```

Comment:
Will a warning suffice? I think it's harmless, the return value differs only by class (list vs. data frame).

Reply:
Let's soft-deprecate.